### DISCUSSION TITLE
Motor function question

### DISCUSSION THREAD

**Discussion by benjamin-grimsey** [pybricks/support#487] [Q&A] [answered]
I have the 42114 Volvo Hauler and I am trying to make a control program for it. I am going through the Pybricks beta coder and also using a 88010 remote. I have the drive and steering working great, but the gearbox has me scratching my head. I want it to  turn +90 degrees incrementally with the red left button, and -90 degrees incrementally with the right red button. I have tried:

```python
    if Button.RIGHT in pressed:
        gearbox.run_angle(150, -90, Stop.HOLD, False)
    elif Button.LEFT in pressed:
        gearbox.run_angle(150, 90, Stop.HOLD, False)
```

and this works, but the motor coasts a little bit farther than +/-90*.

I have also tried:

```python
    if Button.RIGHT in pressed:
        gearbox.run_target(150, -90, Stop.HOLD, False)
    elif Button.LEFT in pressed:
        gearbox.run_target(150, 90, Stop.HOLD, False)
```

and this works, but will only go back and fourth between +90* and -90* with button presses.

I'm quite new to Pybricks so Im not sure how to code it properly.

Thanks for your help!

(Full program is listed below if interested)

```python
from pybricks.pupdevices import Motor, Remote
from pybricks.parameters import Port, Direction, Stop, Button
from pybricks.tools import wait

# Initialize the motors.
front = Motor(Port.A, Direction.COUNTERCLOCKWISE)
gearbox = Motor(Port.B, Direction.COUNTERCLOCKWISE)
steer = Motor(Port.D)

# Connect to the remote.
remote = Remote()

# Read the current settings
old_kp, old_ki, old_kd, _, _ = steer.control.pid()

# Set new values
steer.control.pid(kp=old_kp*4, kd=old_kd*0.4)

# Find the steering endpoint on the left and right.
# The middle is in between.
left_end = steer.run_until_stalled(-200, then=Stop.HOLD)
right_end = steer.run_until_stalled(200, then=Stop.HOLD)

# We are now at the right. Reset this angle to be half the difference.
# That puts zero in the middle.
steer.reset_angle((right_end - left_end)/2)
steer.run_target(speed=200, target_angle=0, wait=False)

# Set steering angle.
steer_angle = (((right_end - left_end)/2)-5)
print('steer angle:',steer_angle)

# Now we can start driving!
while True:
    # Check which buttons are pressed.
    pressed = remote.buttons.pressed()

    # Choose the steer angle based on the right controls.
    if Button.LEFT_PLUS in pressed:
        steer.run_target(1400, -steer_angle, Stop.HOLD, False)
    elif Button.LEFT_MINUS in pressed:
        steer.run_target(1400, steer_angle, Stop.HOLD, False)
    else:
        steer.track_target(0)

    # Choose the drive speed based on the left controls.
    drive_speed = 0
    if Button.RIGHT_PLUS in pressed:
        drive_speed += 100
    if Button.RIGHT_MINUS in pressed:
        drive_speed -= 100

    # Apply the selected speed.
    front.dc(drive_speed)
    
    # Select a gear based on the right and left red controls.
    if Button.RIGHT in pressed:
        gearbox.run_angle(500, -90, Stop.HOLD, False)
    elif Button.LEFT in pressed:
        gearbox.run_angle(500, 90, Stop.HOLD, False)

    # Wait.
    wait(10)

```

**Comment by laurensvalk** [comment, marked as answer]
Both `run_angle` and `run_target` can be used, each in their own way.

`run_angle` is usually great for incremental movement, but you'll want to make sure you start from a known position at the very beginning. So you could use `run_target` to move it to `0` at the start of your program, and then use `run_angle` after that.

But in your case, you use `wait=False`, and you might be giving it a new `run_angle` command when the previous one isn't done yet. That's fine in some applications, but perhaps not in your case. For example, if you are at 45 and you call `run_angle` to add 90 degrees, you're going to 135.

So perhaps you could add a step in between to achieve what you want. Instead of controlling the motors directly with the remote, you could use the buttons of the remote to increment/decrement a variable called `target` by +/-90 degrees. And then use `run_target` to steer towards that `target`. You might also want to add some code to ensure you add it only once each time the button is pressed.

**Reply by benjamin-grimsey** [reply]
Thanks! The majority of this code was copied from one that had already been put together by someone else, and I have been modifying it to work with the gearbox. What you suggest makes sense, I just have no knowledge of how to put that to code, how should I go about learning this? Is it just trial and error?

**Reply by benjamin-grimsey** [reply]
Well, I guess it is trial and error lol. I nixed the "HOLD" part at the end and added a wait command and that seems to do the trick. Thanks!
